features/netsocket/nsapi_dns.cpp

```diff
@@ -305,3 +553,402 @@ nsapi_error_t nsapi_dns_query(NetworkStack *stack, const char *host,
+
+    while (true) {
+        SocketAddress dns_addr;
+        nsapi_size_or_error_t err = nsapi_dns_get_server_addr(query->stack, &(query->dns_server), &(query->total_retries), &dns_addr);
+        if (err != NSAPI_ERROR_OK) {
+            nsapi_dns_query_async_resp(query, NSAPI_ERROR_DNS_FAILURE, NULL);
+            free(packet);
+            dns_mutex.unlock();
+            return;
+        }
+
+        err = query->socket->sendto(dns_addr, packet, len);
+
+        if (err < 0) {
+            query->dns_server++;
+        } else {
+            break;
+        }
+    }
+
+    if (query->total_retries) {
+        query->total_retries--;
+    }
+
+    free(packet);
+
+    if (nsapi_dns_call_in(query->call_in_cb, MBED_CONF_NSAPI_DNS_RESPONSE_WAIT_TIME,
+        mbed::callback(nsapi_dns_query_async_send, reinterpret_cast<void *>(unique_id))) != NSAPI_ERROR_OK) {
+        nsapi_dns_query_async_resp(query, NSAPI_ERROR_NO_MEMORY, NULL);
+    }
+
+    dns_mutex.unlock();
+}
+
+static void nsapi_dns_query_async_socket_callback(NetworkStack *stack)
+{
+    UDPSocket *socket = NULL;
+
+    dns_mutex.lock();
+
+    for (int i = 0; i < DNS_QUERY_QUEUE_SIZE; i++) {
+        if (dns_query_queue[i] && dns_query_queue[i]->stack == stack) {
+            socket = dns_query_queue[i]->socket;
+            break;
+        }
+    }
+
+    if (socket) {
+        // create network packet
+        uint8_t *packet = (uint8_t *)malloc(DNS_BUFFER_SIZE);
+        if (!packet) {
+            dns_mutex.unlock();
+            return;
+        }
+
+        while (true) {
+            // recv the response
+            nsapi_size_or_error_t size = socket->recvfrom(NULL, packet, DNS_BUFFER_SIZE);
+
+            if (size < DNS_RESPONSE_MIN_SIZE) {
+                break;
+            }
+
+            // gets id from response to associate with correct query
+            uint16_t id = (packet[0] << 8) | packet[1];
+
+            DNS_QUERY *query = NULL;
+
+            for (int i = 0; i < DNS_QUERY_QUEUE_SIZE; i++) {
+                if (dns_query_queue[i] && dns_query_queue[i]->dns_message_id == id) {
+                    query = dns_query_queue[i];
+                    break;
+                }
+            }
+
+            if (!query) {
+                continue;
+            }
+
+            int requested_count = 1;
+            if (query->addr_count > 1) {
+                requested_count = query->addr_count;
+            }
+
+            query->addrs = new (std::nothrow) nsapi_addr_t[requested_count];
+
+            int resp = dns_scan_response(packet, id, &(query->ttl), query->addrs, requested_count);
+
+            // Ignore invalid responses
+            if (resp < 0) {
+                delete[] query->addrs;
+                query->addrs = 0;
+            } else {
+                query->count = resp;
+                nsapi_dns_call_in(query->call_in_cb, 0, mbed::callback(nsapi_dns_query_async_response, reinterpret_cast<void *>(query->unique_id)));
```

Comment:
Is this cast necessary?

Reply:
Yes, seems that adding an int to callback is not possible.